gcc 7.2 compilation fails

### Description of the enhancement or error report

MOOSE framework compilation fails for me with
```
In file included from /h2/roystgnr/git/moose/modules/tensor_mechanics/include/materials/CompositeEigenstrain.h:11:0,
                 from /h2/roystgnr/git/moose/modules/tensor_mechanics/src/base/TensorMechanicsApp.C:57:
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h: In constructor ‘CompositeTensorBase<T, U>::CompositeTensorBase(const InputParameters&)’:
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h:83:60: error: expected ‘)’ before ‘::’ token
     _num_args(this->template DerivativeMaterialInterface<U>::coupledComponents("args")),
                                                            ^~
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h: In member function ‘void CompositeTensorBase<T, U>::initializeDerivativeProperties(std::__cxx11::string)’:
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h:117:54: error: expected ‘,’ or ‘;’ before ‘::’ token
         this->template DerivativeMaterialInterface<U>::getVar("args", j)->name();
                                                      ^~
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h:124:56: error: expected ‘,’ or ‘;’ before ‘::’ token
           this->template DerivativeMaterialInterface<U>::getVar("args", k)->name();
                                                        ^~
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h:143:56: error: expected ‘,’ or ‘;’ before ‘::’ token
           this->template DerivativeMaterialInterface<U>::getVar("args", j)->name();
                                                        ^~
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h:156:58: error: expected ‘,’ or ‘;’ before ‘::’ token
             this->template DerivativeMaterialInterface<U>::getVar("args", k)->name();
                                                          ^~
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h: In member function ‘virtual void CompositeTensorBase<T, U>::computeQpTensorProperties(MaterialProperty<P>&, libMesh::Real)’:
/h2/roystgnr/git/moose/framework/include/materials/CompositeTensorBase.h:172:72: error: expected ‘,’ or ‘;’ before ‘::’ token
   const unsigned int qp = this->template DerivativeMaterialInterface<U>::_qp;
```

### Rationale for the enhancement or information for reproducing the error

Switching to gcc 7.2 instead of 6.1 triggered the error for me; I believe that would be sufficient for others too.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Looking through C++ standards documents, it *appears* to me that the template keyword here, referring to a superclass, shouldn't be needed, because the superclass isn't a member template in need of a qualifier... but I and gcc 7 may be misunderstanding that (in which case this is not a bug, and "fixing" it would actually introduce a bug to other compilers) or other compilers may misunderstand that (in which case this code is a bug, but fixing it would break a workaround for other compilers).

--- a/framework/include/materials/CompositeTensorBase.h
+++ b/framework/include/materials/CompositeTensorBase.h
@@ -80,7 +80,7 @@ CompositeTensorBase<T, U>::CompositeTensorBase(const InputParameters & parameter
   : DerivativeMaterialInterface<U>(parameters),
     _tensor_names(this->template getParam<std::vector<MaterialPropertyName>>("tensors")),
     _weight_names(this->template getParam<std::vector<MaterialPropertyName>>("weights")),
-    _num_args(this->template DerivativeMaterialInterface<U>::coupledComponents("args")),
+    _num_args(this->DerivativeMaterialInterface<U>::coupledComponents("args")),
     _num_comp(_tensor_names.size()),
     _dM(_num_args),
     _d2M(_num_args),
@@ -114,14 +114,14 @@ CompositeTensorBase<T, U>::initializeDerivativeProperties(const std::string name
   for (unsigned int j = 0; j < _num_args; ++j)
   {
     const VariableName & jname =
-        this->template DerivativeMaterialInterface<U>::getVar("args", j)->name();
+        this->DerivativeMaterialInterface<U>::getVar("args", j)->name();
     _dM[j] = &this->template declarePropertyDerivative<T>(name, jname);
     _d2M[j].resize(j + 1);
 
     for (unsigned int k = 0; k <= j; ++k)
     {
       const VariableName & kname =
-          this->template DerivativeMaterialInterface<U>::getVar("args", k)->name();
+          this->DerivativeMaterialInterface<U>::getVar("args", k)->name();
       _d2M[j][k] = &this->template declarePropertyDerivative<T>(name, jname, kname);
     }
   }
@@ -140,7 +140,7 @@ CompositeTensorBase<T, U>::initializeDerivativeProperties(const std::string name
     for (unsigned int j = 0; j < _num_args; ++j)
     {
       const VariableName & jname =
-          this->template DerivativeMaterialInterface<U>::getVar("args", j)->name();
+          this->DerivativeMaterialInterface<U>::getVar("args", j)->name();
 
       _dtensors[i][j] =
           &this->template getMaterialPropertyDerivativeByName<T>(_tensor_names[i], jname);
@@ -153,7 +153,7 @@ CompositeTensorBase<T, U>::initializeDerivativeProperties(const std::string name
       for (unsigned int k = 0; k <= j; ++k)
       {
         const VariableName & kname =
-            this->template DerivativeMaterialInterface<U>::getVar("args", k)->name();
+            this->DerivativeMaterialInterface<U>::getVar("args", k)->name();
 
         _d2tensors[i][j][k] =
             &this->template getMaterialPropertyDerivativeByName<T>(_tensor_names[i], jname, kname);
@@ -169,7 +169,7 @@ void
 CompositeTensorBase<T, U>::computeQpTensorProperties(MaterialProperty<T> & M,
                                                      Real derivative_prefactor)
 {
-  const unsigned int qp = this->template DerivativeMaterialInterface<U>::_qp;
+  const unsigned int qp = this->DerivativeMaterialInterface<U>::_qp;
 
   M[qp].zero();
   for (unsigned int i = 0; i < _num_comp; ++i)

--- a/framework/include/materials/CompositeTensorBase.h
+++ b/framework/include/materials/CompositeTensorBase.h
@@ -113,15 +113,13 @@ CompositeTensorBase<T, U>::initializeDerivativeProperties(const std::string name
   // setup output composite tensor and derivatives
   for (unsigned int j = 0; j < _num_args; ++j)
   {
-    const VariableName & jname =
-        this->DerivativeMaterialInterface<U>::getVar("args", j)->name();
+    const VariableName & jname = this->DerivativeMaterialInterface<U>::getVar("args", j)->name();
     _dM[j] = &this->template declarePropertyDerivative<T>(name, jname);
     _d2M[j].resize(j + 1);
 
     for (unsigned int k = 0; k <= j; ++k)
     {
-      const VariableName & kname =
-          this->DerivativeMaterialInterface<U>::getVar("args", k)->name();
+      const VariableName & kname = this->DerivativeMaterialInterface<U>::getVar("args", k)->name();
       _d2M[j][k] = &this->template declarePropertyDerivative<T>(name, jname, kname);
     }
   }
@@ -139,8 +137,7 @@ CompositeTensorBase<T, U>::initializeDerivativeProperties(const std::string name
 
     for (unsigned int j = 0; j < _num_args; ++j)
     {
-      const VariableName & jname =
-          this->DerivativeMaterialInterface<U>::getVar("args", j)->name();
+      const VariableName & jname = this->DerivativeMaterialInterface<U>::getVar("args", j)->name();
 
       _dtensors[i][j] =
           &this->template getMaterialPropertyDerivativeByName<T>(_tensor_names[i], jname);

